fix QApplication initialization and allow multiple instances of vizkit3d_world

alcantara09 opened the pull request.

Commit:
add missing dependency

Commit:
fix QApplication intialization and deletion

QApplication requires that the data referred to by argc and argv must
stay valid for the entire lifetime of the QApplication object. In
addition, argc must be greater than zero and argv must contain at least
one valid character string.

Besides this it should not be responsability of the class to spawn the
QApplication, that must be a singleton. Rock already has support to tag
if a component uses qt and this will be utilized in future
implementations. Backward compatibility is assured if someone tries to
run vizkit3dworld withou spawning the application beforehand.

Commit:
remove copy construction creation

Commit:
add unit test for multiple instances

renan028 commented on src/Vizkit3dWorld.cpp:
```diff
@@ -20,6 +20,9 @@
 namespace vizkit3d_world {
 
+static int QT_ARGC = 1;
+static char* QT_ARGV[1] = { "vizkit3d_world" };
```
warning: ISO C++ forbids converting a string constant to ‘char*’

```suggestion
static char* QT_ARGV[1] = {(char *)"vizkit3d_world"};
```

doudou commented on src/Vizkit3dWorld.cpp:
```diff
@@ -36,12 +39,14 @@ Vizkit3dWorld::Vizkit3dWorld(std::string path,
-    int argc = 1;
-    char const*argv[] = { "vizkit3d_world" };
-    app = new QApplication(argc, const_cast<char**>(argv));
+        app = new QApplication(QT_ARGC, const_cast<char**>(QT_ARGV));
```
Even if, technically, removing the `app` attribute would break backward compatibility, I think it would be best to remove the attribute, since after the first initialized vizkit3dworld object, it *will* be NULL. It would ensure that subclasses that use `app` will not compile anymore (instead of crashing when creating their second Vizkit3dWorld object).

Commit:
Merge branch 'master' into multiple_vizkit3dworld

This is merge is necessary because we are dropping all the changes
related to multiple widget. Multithreading is going to be performed
using multiple Vizkit3dWorld instances

Commit:
remove Qapplication pointer

Even if, technically, removing the app attribute would break backward
compatibility, I think it would be best to remove the attribute, since
after the first initialized vizkit3dworld object, it will be NULL. It
would ensure that subclasses that use app will not compile anymore
(instead of crashing when creating their second Vizkit3dWorld object).

renan028 reviewed (commented):
I am getting the following error when I test:
./build/test/test_suite --run_test=it_can_create_multiple_vizkit3_world
Running 1 test case...
QWidget: Must construct a QApplication before a QPaintDevice
unknown location(0): fatal error: in "it_can_create_multiple_vizkit3_world": signal: SIGABRT (application abort requested)
/home/renan/dev/squid/gui/vizkit3d_world/test/testVizkit3dWorld.cpp(15): last checkpoint: "it_can_create_multiple_vizkit3_world" test entry

*** 1 failure is detected in the test module "Master Test Suite"

renan028 commented on test/testVizkit3dWorld.cpp:
```diff
@@ -13,17 +12,14 @@ BOOST_AUTO_TEST_CASE(it_loads_correctly_a_sdf_world)
-        2);
-    BOOST_TEST(vizkit3dWorld->getWidget(0)->isWindow(),1);
-    BOOST_TEST(vizkit3dWorld->getWidget(1)->isWindow(),1);
+    vizkit3d_world::Vizkit3dWorld vizkit3d_world("test_data/simple.world");
```
Initialize the app before:
```suggestion
    QApplication* app;
    int argc = 1;
    char const*argv[] = { (char *)"vizkit3d_world" };
    app = new QApplication(argc, const_cast<char**>(argv));
    vizkit3d_world::Vizkit3dWorld vizkit3d_world("test_data/simple.world");
```

renan028 commented on test/testVizkit3dWorld.cpp:
```diff
@@ -13,17 +12,14 @@ BOOST_AUTO_TEST_CASE(it_loads_correctly_a_sdf_world)
     BOOST_CHECK_EQUAL(robot_viz.begin()->first,"simple_model");
```
Initialize the app before calling vizkit3d_world:
   QApplication* app;
    int argc = 1;
    char const*argv[] = { (char *)"vizkit3d_world" };
    app = new QApplication(argc, const_cast<char**>(argv));

renan028 commented on src/Vizkit3dWorld.cpp:
```diff
@@ -20,46 +20,45 @@
-    char const*argv[] = { "vizkit3d_world" };
-    app = new QApplication(argc, const_cast<char**>(argv));
+        QApplication app(QT_ARGC, const_cast<char**>(QT_ARGV));
+    }
```
the app does not exist outside the "if" scope.
If you call:
std::cout << QApplication::instance() << std:endl;

it returns 0.
If you make a pointer: 
QApplication *app = QApplication(QT_ARGC, const_cast<char**>(QT_ARGV));
it works, but then you will have to delete it somehow.

alcantara09 replied:
The init and responsability is blurry here. Vizkit3dWorld should not have the ownership over QApplication. However the use_qt method in task_context is spawninng a QApplication in a different thread and it is complaining about it. For now we are going to let the QApplication leak and open an issue.

Commit:
fix initialization of QApplication and unit tests
